src/config/secret.rs

```diff
@@ -0,0 +1,241 @@
+use std::collections::HashMap;
+
+use futures::executor;
+use indexmap::IndexMap;
+use regex::{Captures, Regex};
+use serde::{Deserialize, Serialize};
+use tokio::{io::AsyncWriteExt, process::Command, time::Duration};
+use typetag::serde;
+
+use super::{format, ComponentKey, Format};
+
+#[typetag::serde(tag = "type")]
+pub trait SecretBackend: core::fmt::Debug + Send + Sync + dyn_clone::DynClone {
+    fn retrieve(&mut self, secret_keys: Vec<String>) -> crate::Result<HashMap<String, String>>;
+}
+
+#[derive(Deserialize, Serialize, Debug, Default)]
+struct SecretBackends {
+    #[serde(default)]
+    secret: IndexMap<ComponentKey, Box<dyn SecretBackend>>,
+}
+
+pub fn interpolate(input: &str, format: Format) -> (String, Vec<String>) {
+    let keys = collect_secret_keys(input);
+    if keys.is_empty() {
+        debug!("No secret placeholder found, skipping secret resolution.");
+        return (input.to_owned(), Vec::new());
+    }
+    let backends = format::deserialize::<SecretBackends>(input, format);
+    match backends {
+        Err(e) => (input.to_owned(), e),
+        Ok(backends) => {
+            let (secrets, warnings) = retrieve_secrets(keys, backends);
+            (do_replace(input, secrets), warnings)
+        }
+    }
+}
+
+fn retrieve_secrets(
+    keys: HashMap<String, Vec<String>>,
+    mut backends: SecretBackends,
+) -> (HashMap<String, String>, Vec<String>) {
+    let mut warnings = Vec::<String>::new();
+    let secrets = keys.into_iter().flat_map(|(backend_name, keys)| {
+        match backends.secret.get_mut(&ComponentKey::from(backend_name.clone())) {
+            None => {
+                warnings.push(format!("Backend \"{}\" is required for secret retrieval but was not found in config.", backend_name));
+                vec![]
+            },
+            Some(backend) => {
+                match backend.retrieve(keys) {
+                    Err(e) => {
+                        warnings.push(e.to_string());
+                        vec![]
+                    },
+                    Ok(s) => {
+                        s.into_iter().map(|(k, v)| {
+                            (format!("{}.{}", backend_name, k), v)
+                        }).collect::<Vec<(String, String)>>()
+                    }
+                }
+            },
+        }
+    }).collect::<HashMap<String,String>>();
+    (secrets, warnings)
+}
+
+fn do_replace(input: &str, secrets: HashMap<String, String>) -> String {
+    let re = Regex::new(r"SECRET\[([[:word:]]+\.[[:word:].]+)\]").unwrap();
```

Comment:
We've been doing this in a [`once_cell::sync::Lazy`](https://docs.rs/once_cell/latest/once_cell/sync/struct.Lazy.html), should we do that here as well? It's obviously not in a hot path, but perhaps for consistency's sake?

Reply:
Yeah, agreed, we should pull this up as a lazy value for consistency. I realize this function is only called once right now, but it definitely jumps out when scanning.

Reply:
Will do 👍